features/steps/steps.py

```diff
@@ -285,6 +382,71 @@ def step_impl(context, constraint, num, entity):
 
     handle_errors(context, errors)
 
+def get_duplicates(values):
+    seen = set()
+    duplicates = [x for x in values if x in seen or seen.add(x)]
+    return duplicates
+
+def evaluate_identical_unique(msg, stack_tree, i, identical_or_unique, relating):
+    if (
+        identical_or_unique == 'identical' and
+        len(msg.values) > 1 and
+        not msg.duplicates
+    ):
+        return msg.values, relating, stack_tree[-1] # values, relating, related
+
+    elif(
+        identical_or_unique == 'unique' and
+        len(msg.duplicates)
+    ):  
+        inst_tree = [t[i] for t in stack_tree]
+        false_instances = [inst_tree[1][i] for i, x in enumerate(msg.values) if x in msg.duplicates]
+        return msg.duplicates, false_instances, inst_tree[-1] # values, relating, related
+
+    else: return None, None, None
+
+def convert_values(values, context):
+    """
+    Converts ifcopenshell instance type to strings to check for duplicates, if applicable
+    Perhaps also specify output type (entity instances/integers/strings etc) for further error analysis
+    """
+    converted_values = []
+    for value in values:
+        try:
+            converted_values.append(value[0].is_a())
+            setattr(context, 'include_relating_entities', False)
+        except (IndexError, AttributeError):
+            value = 'None' if value == () else value
+            converted_values.append(value)
+    return converted_values
+
+@then("The values must be {identical_or_unique}")
+def step_impl(context, identical_or_unique):
+    errors = []
+
+    within_model = getattr(context, 'within_model', False)
+
+    if getattr(context, 'applicable', True):
+        stack_tree = list(filter(None, list(map(lambda layer: layer.get('instances'), context._stack))))
+        instances = [context.instances] if within_model else context.instances
+
+        for i, values in enumerate(instances):
+            if not values:
+                continue
+
+            msg = value_error_msg(identical_or_unique=identical_or_unique, attribute=getattr(context, 'attribute', 'None'))
```

Comment:
I'm a bit confused by all of this. You need to help me a bit:

- You overwrite `msg.values`.
- I would prefer to have the `value_error_msg` class to be initialized completely. Now you have everything with default values as "None". But they are not really optional you just supply them later. This is confusing to the reader. Rather use local variables here, and initialize the error class entirely in a single call.
- I'd always expect a function named `convert_values()` to be "pure", i.e no side effects, but this function sets include_relating_entities on the context which is confusing. It's also unclear what this triggers because there are no comments in this section.
- convert_values could also be named a bit more descriptive, I'd imagine in other steps we'll be doing also a lot of 
converting.
- the error class is also very complex, maybe we've reached a point that it needs to be broken down into several classes. I think the formatting within them should be purely obvious.

Reply:
It seems initializing the dataclass first and supplying data later leads to more confusion than clarity. It's changed in https://github.com/buildingSMART/ifc-gherkin-rules/commit/a3c7871507f07fbcb83cc95d839259780d7f4b7e as well as https://github.com/buildingSMART/ifc-gherkin-rules/commit/6ba2c0a43788bdefacf2905347e7853849d02814

The dataclass is also simplified, and the logic is moved to the 'Then' statements that use this dataclass. This, I think, makes it more obvious to read.